File: fromchatgpt/claude_code_fhr_project_bundle/code/python/02_data_cleaning.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Data cleaning and variable construction for SSJDA 1331 analysis.
Constructs all variables needed for reproducible analysis.
Output saved to data/derived/ for use in subsequent analysis.
"""

import pandas as pd
import numpy as np
from pathlib import Path

# Setup paths
PROJECT_ROOT = Path(__file__).parent.parent.parent
AUTHORITATIVE_DATA_DIR = Path(r"C:\Users\yuyaa\git\postwar_household_insurance\postwar_household_insurance\data\raw")
SSJDA_EXTRACT_DIR = AUTHORITATIVE_DATA_DIR / "神奈川県における民生基礎調査（ボーダー・ライン層調査）1961" / "1331"
SSJDA_CSV_FILE = SSJDA_EXTRACT_DIR / "1331.csv"
DATA_DERIVED = PROJECT_ROOT / "data" / "derived"
DATA_DERIVED.mkdir(parents=True, exist_ok=True)

print("=" * 80)
print("DATA CLEANING AND VARIABLE CONSTRUCTION")
print("=" * 80)

# Load raw data
print("\n1. Loading raw data...")
df = pd.read_csv(SSJDA_CSV_FILE)
print(f"   Loaded: {len(df):,} observations")

# Keep original for reference
df_orig = df.copy()

# ============================================================================
# INSTITUTIONAL VARIABLES
# ============================================================================

print("\n2. Constructing institutional variables...")

# Public program use (binary)
df['any_welfare_loan'] = ((df['q04_2'] == 1) | (df['q04_3'] == 1) |
                          (df['q04_4'] == 1) | (df['q04_7'] == 1)).astype(int)
df['public_pawnshop'] = (df['q04_8'] == 1).astype(int)

# Institutional groups
df['both_institutions'] = ((df['any_welfare_loan'] == 1) &
                           (df['public_pawnshop'] == 1)).astype(int)
df['welfare_only'] = ((df['any_welfare_loan'] == 1) &
                      (df['public_pawnshop'] == 0)).astype(int)
df['pawnshop_only'] = ((df['any_welfare_loan'] == 0) &
                       (df['public_pawnshop'] == 1)).astype(int)
df['neither_institution'] = ((df['any_welfare_loan'] == 0) &
                             (df['public_pawnshop'] == 0)).astype(int)

# Public assistance (for control variables)
df['public_assistance'] = (df['q04_1'] == 1).astype(int)

print("   Welfare loan: {:,}".format(df['any_welfare_loan'].sum()))
print("   Pawnshop: {:,}".format(df['public_pawnshop'].sum()))
print("   Both: {:,}".format(df['both_institutions'].sum()))
print("   Welfare only: {:,}".format(df['welfare_only'].sum()))
print("   Pawnshop only: {:,}".format(df['pawnshop_only'].sum()))
print("   Neither: {:,}".format(df['neither_institution'].sum()))

# ============================================================================
# COPING STRATEGIES (Q10 VARIABLES)
# ============================================================================

print("\n3. Constructing coping strategy variables...")

coping_vars = {
    'coping_account': 'q10_1',          # Purchases on account
    'coping_pawn': 'q10_2',             # Pawning
    'coping_employer': 'q10_3',         # Employer borrowing
    'coping_friend': 'q10_4',           # Friend/neighbor borrowing
    'coping_asset_sale': 'q10_5',       # Asset sales
    'coping_savings': 'q10_6',          # Savings withdrawal
    'coping_other': 'q10_7',            # Other
    'coping_food': 'q10_8'              # Food compression
}

for new_var, orig_var in coping_vars.items():
    df[new_var] = (df[orig_var] == 1).astype(int)

# Count of coping strategies
df['n_coping_strategies'] = df[[v for v in coping_vars.keys()]].sum(axis=1)
df['any_coping'] = (df['n_coping_strategies'] > 0).astype(int)

# Borrowing indicators (trade credit, relational, collateral)
df['any_borrowing'] = ((df['q10_1'] == 1) | (df['q10_2'] == 1) |
                       (df['q10_3'] == 1) | (df['q10_4'] == 1)).astype(int)

print("   Coping strategy variables created")
print("   Mean number of strategies: {:.2f}".format(df['n_coping_strategies'].mean()))

# ============================================================================
# HOUSEHOLD DEMOGRAPHICS
# ============================================================================

print("\n4. Constructing demographic variables...")

# Head age (from q01_01_3, with 99 = missing)
df['head_age'] = df['q01_01_3'].where(df['q01_01_3'] < 99, np.nan)
df['head_age_sq'] = df['head_age'] ** 2

# Female head (1=female, 0=male)
df['female_head'] = (df['q01_01_2'] == 2).astype(int)
# Set to NaN if not properly coded
df.loc[~df['q01_01_2'].isin([1, 2]), 'female_head'] = np.nan

# Household size (from q12_1, with 99 = missing)
df['household_size'] = df['q12_1'].where(df['q12_1'] < 99, np.nan)

# Workers in household (from q12_4, with 9 = missing)
df['n_workers'] = df['q12_4'].where(df['q12_4'] < 9, np.nan)

# Unemployed in household (from q12_6, with 9 = missing)
df['n_unemployed'] = df['q12_6'].where(df['q12_6'] < 9, np.nan)

print("   Head age - mean: {:.1f}".format(df['head_age'].mean()))
print("   Female head: {:.1%}".format(df['female_head'].mean()))
print("   Household size - mean: {:.1f}".format(df['household_size'].mean()))
print("   Workers - mean: {:.1f}".format(df['n_workers'].mean()))

# ============================================================================
# HOUSEHOLD ASSETS (Q26 VARIABLES)
# ============================================================================

print("\n5. Constructing asset variables...")

# Asset ownership (0 = not owned, 1 = owned, 9 = missing/unknown)
asset_items = ['q26_{}'.format(i) for i in range(1, 10)]
asset_labels = [
    'dresser', 'sewing_machine', 'radio', 'television',
    'watch', 'dining_table', 'electric_washer', 'study_desk', 'real_estate'
]

# Clean assets (set 9 = NaN)
for var in asset_items:
    df[var] = df[var].where(df[var] != 9, np.nan)

# Asset count (only owned=1, exclude real estate first)
df['asset_count'] = df[[f'q26_{i}' for i in range(1, 9)]].sum(axis=1)
df['asset_count_with_realestate'] = df[asset_items].sum(axis=1)

print("   Asset count (8 movables) - mean: {:.2f}".format(df['asset_count'].mean()))
print("   Asset count (including real estate) - mean: {:.2f}".format(df['asset_count_with_realestate'].mean()))

# ============================================================================
# HOUSEHOLD ECONOMIC CHARACTERISTICS (Q02 VARIABLES)
# ============================================================================

print("\n6. Constructing low-income-cause variables...")

# Q02 variables: risk factors for low income (0=no, 1=yes, 99=missing)
# We use the coding structure from the preliminary analysis

risk_factors = {
    'war_damage': 'q02_01',
    'disaster': 'q02_02',
    'head_death': 'q02_03',
    'unemployment': 'q02_04',
    'business_failure': 'q02_05',
    'family_conflict': 'q02_06',
    'income_decline': 'q02_07',
    'asset_loss': 'q02_08',
    'illness_onset': 'q02_09',
    'prolonged_illness': 'q02_10',
    'aging_work_decline': 'q02_11',
    'low_living_ability': 'q02_21',
    'weak_household_head': 'q02_22',
    'household_discord': 'q02_23',
    'disabled_household_member': 'q02_24',
    'longterm_patient': 'q02_25'
}

for new_var, orig_var in risk_factors.items():
    # Convert to binary, set 99 (missing/unknown) to NaN
    df[new_var] = df[orig_var].where(df[orig_var] != 99, np.nan)

print("   Low-income-cause variables created: {}".format(len(risk_factors)))

# ============================================================================
# HOUSEHOLD INCOME AND CONSUMPTION
# ============================================================================

print("\n7. Constructing income and consumption variables...")

# Income level (q07): categorical, 1-16 (higher = more income)
# Keep as is; will use as categorical in models

# Income change (q08): 1=increased, 2=no change, 3=decreased, 9=missing
df['income_decreased'] = (df['q08'] == 3).astype(int)
df.loc[df['q08'] == 9, 'income_decreased'] = np.nan

# Rent arrears (q19): amount in yen, 88888 or >1000 = missing/not applicable
df['rent_arrears_amount'] = df['q19'].where((df['q19'] < 1000) & (df['q19'] != 88888), np.nan)
df['has_rent_arrears'] = (df['rent_arrears_amount'] > 0).astype(int)

print("   Income decreased: {:.1%}".format(df['income_decreased'].mean()))
print("   Has rent arrears: {:.1%}".format(df['has_rent_arrears'].mean()))

# ============================================================================
# HOUSEHOLD TYPE
# ============================================================================

print("\n8. Household type and employment...")

# Household type (q03): keep as is for now
df['household_type'] = df['q03']

print("   Household types: {}".format(df['household_type'].nunique()))

# ============================================================================
# SAVE ANALYSIS DATASET
# ============================================================================

print("\n9. Saving analysis dataset...")

# Select key variables for analysis
key_columns = ['ID', 'city', 'household_type'] + \
              list(coping_vars.keys()) + \
              ['any_coping', 'n_coping_strategies', 'any_borrowing'] + \
              ['any_welfare_loan', 'public_pawnshop', 'public_assistance'] + \
              ['both_institutions', 'welfare_only', 'pawnshop_only', 'neither_institution'] + \
              ['head_age', 'head_age_sq', 'female_head', 'household_size',
               'n_workers', 'n_unemployed'] + \
              ['asset_count', 'asset_count_with_realestate'] + \
              list(risk_factors.keys()) + \
              ['income_decreased', 'has_rent_arrears', 'rent_arrears_amount'] + \
              ['q04_1', 'q04_2', 'q04_3', 'q04_4', 'q04_5', 'q04_6', 'q04_7', 'q04_8',  # Original Q04
               'q07', 'q08', 'q10_1', 'q10_2', 'q10_3', 'q10_4', 'q10_5', 'q10_6', 'q10_7', 'q10_8']  # Original Q10

# Create analysis dataset
df_analysis = df[key_columns].copy()

# Save to CSV
output_file = DATA_DERIVED / "ssjda1331_analysis.csv"
df_analysis.to_csv(output_file, index=False)

print(f"   Saved to: {output_file}")
print(f"   Dimensions: {df_analysis.shape}")

# ============================================================================
# DATA QUALITY SUMMARY
# ============================================================================

print("\n10. Data quality summary...")

print("\n   Missing values (key institutional variables):")
print(f"     any_welfare_loan: {df_analysis['any_welfare_loan'].isna().sum()}")
print(f"     public_pawnshop: {df_analysis['public_pawnshop'].isna().sum()}")

print("\n   Missing values (key demographic variables):")
print(f"     head_age: {df_analysis['head_age'].isna().sum()}")
print(f"     female_head: {df_analysis['female_head'].isna().sum()}")
print(f"     n_workers: {df_analysis['n_workers'].isna().sum()}")

print("\n   Coping variables complete: YES (all binary for full sample)")

print("\n" + "=" * 80)
print("DATA CLEANING COMPLETE")
print("=" * 80)
print(f"Output file: {output_file}")
print(f"Ready for regression analysis.")

File: fromchatgpt/claude_code_fhr_project_bundle/code/02_data_cleaning.R
# Data cleaning and variable construction for SSJDA 1331 analysis
# All outputs saved to data/derived/ for use in subsequent analysis

library(tidyverse)

# Setup paths directly
project_root <- "C:/Users/yuyaa/git/postwar_household_insurance/postwar_household_insurance/fromchatgpt/claude_code_fhr_project_bundle"
authoritative_data_dir <- "C:/Users/yuyaa/git/postwar_household_insurance/postwar_household_insurance/data/raw"
ssjda_extract_dir <- file.path(
  authoritative_data_dir,
  "神奈川県における民生基礎調査（ボーダー・ライン層調査）1961",
  "1331"
)
ssjda_csv_file <- file.path(ssjda_extract_dir, "1331.csv")
data_derived <- file.path(project_root, "data", "derived")
output_logs <- file.path(project_root, "output", "logs")
dir.create(data_derived, recursive = TRUE, showWarnings = FALSE)
dir.create(output_logs, recursive = TRUE, showWarnings = FALSE)

cat("================================================================================\n")
cat("DATA CLEANING AND VARIABLE CONSTRUCTION\n")
cat("================================================================================\n")

# ============================================================================
# 1. LOAD DATA
# ============================================================================

cat("\n1. Loading raw data...\n")
df <- read_csv(ssjda_csv_file, show_col_types = FALSE)
cat(sprintf("   Loaded: %s observations\n", format(nrow(df), big.mark = ",")))

# ============================================================================
# 2. INSTITUTIONAL VARIABLES
# ============================================================================

cat("\n2. Constructing institutional variables...\n")

df <- df %>%
  mutate(
    any_welfare_loan = as.integer((q04_2 == 1 | q04_3 == 1 | q04_4 == 1 | q04_7 == 1)),
    public_pawnshop = as.integer(q04_8 == 1),
    both_institutions = as.integer(any_welfare_loan == 1 & public_pawnshop == 1),
    welfare_only = as.integer(any_welfare_loan == 1 & public_pawnshop == 0),
    pawnshop_only = as.integer(any_welfare_loan == 0 & public_pawnshop == 1),
    neither_institution = as.integer(any_welfare_loan == 0 & public_pawnshop == 0),
    public_assistance = as.integer(q04_1 == 1)
  )

cat(sprintf("   Welfare loan: %s\n", format(sum(df$any_welfare_loan), big.mark = ",")))
cat(sprintf("   Pawnshop: %s\n", format(sum(df$public_pawnshop), big.mark = ",")))
cat(sprintf("   Both: %s\n", format(sum(df$both_institutions), big.mark = ",")))
cat(sprintf("   Welfare only: %s\n", format(sum(df$welfare_only), big.mark = ",")))
cat(sprintf("   Pawnshop only: %s\n", format(sum(df$pawnshop_only), big.mark = ",")))
cat(sprintf("   Neither: %s\n", format(sum(df$neither_institution), big.mark = ",")))

# ============================================================================
# 3. COPING STRATEGIES (Q10 VARIABLES)
# ============================================================================

cat("\n3. Constructing coping strategy variables...\n")

df <- df %>%
  mutate(
    coping_account = as.integer(q10_1 == 1),
    coping_pawn = as.integer(q10_2 == 1),
    coping_employer = as.integer(q10_3 == 1),
    coping_friend = as.integer(q10_4 == 1),
    coping_asset_sale = as.integer(q10_5 == 1),
    coping_savings = as.integer(q10_6 == 1),
    coping_other = as.integer(q10_7 == 1),
    coping_food = as.integer(q10_8 == 1)
  ) %>%
  mutate(
    n_coping_strategies = coping_account + coping_pawn + coping_employer +
                         coping_friend + coping_asset_sale + coping_savings +
                         coping_other + coping_food,
    any_coping = as.integer(n_coping_strategies > 0),
    any_borrowing = as.integer((q10_1 == 1 | q10_2 == 1 | q10_3 == 1 | q10_4 == 1))
  )

cat(sprintf("   Mean number of strategies: %.2f\n", mean(df$n_coping_strategies)))

# ============================================================================
# 4. HOUSEHOLD DEMOGRAPHICS
# ============================================================================

cat("\n4. Constructing demographic variables...\n")

df <- df %>%
  mutate(
    head_age = if_else(q01_01_3 < 99, q01_01_3, NA_integer_),
    head_age_sq = head_age ^ 2,
    female_head = if_else(q01_01_2 == 2, 1L, if_else(q01_01_2 == 1, 0L, NA_integer_)),
    household_size = if_else(q12_1 < 99, q12_1, NA_integer_),
    n_workers = if_else(q12_4 < 9, q12_4, NA_integer_),
    n_unemployed = if_else(q12_6 < 9, q12_6, NA_integer_)
  )

cat(sprintf("   Head age - mean: %.1f\n", mean(df$head_age, na.rm = TRUE)))
cat(sprintf("   Female head: %.1f%%\n", 100 * mean(df$female_head, na.rm = TRUE)))
cat(sprintf("   Household size - mean: %.1f\n", mean(df$household_size, na.rm = TRUE)))
cat(sprintf("   Workers - mean: %.1f\n", mean(df$n_workers, na.rm = TRUE)))

# ============================================================================
# 5. HOUSEHOLD ASSETS (Q26 VARIABLES)
# ============================================================================

cat("\n5. Constructing asset variables...\n")

# Asset ownership: recode 9 = NaN
for (i in 1:9) {
  var_name <- paste0("q26_", i)
  df[[var_name]] <- if_else(df[[var_name]] == 9, NA_integer_, df[[var_name]])
}

# Asset count (8 movables, excluding real estate)
df <- df %>%
  mutate(
    asset_count = rowSums(select(., q26_1:q26_8), na.rm = TRUE),
    asset_count_with_realestate = rowSums(select(., q26_1:q26_9), na.rm = TRUE)
  )

cat(sprintf("   Asset count (8 movables) - mean: %.2f\n", mean(df$asset_count, na.rm = TRUE)))
cat(sprintf("   Asset count (including real estate) - mean: %.2f\n", mean(df$asset_count_with_realestate, na.rm = TRUE)))

# ============================================================================
# 6. LOW-INCOME-CAUSE VARIABLES (Q02)
# ============================================================================

cat("\n6. Constructing low-income-cause variables...\n")

# Q02 variables: recode 99 = NaN
q02_vars <- paste0("q02_", c(sprintf("%02d", 1:11), sprintf("%02d", 21:25)))

for (var in q02_vars) {
  if (var %in% names(df)) {
    df[[var]] <- if_else(df[[var]] == 99, NA_integer_, df[[var]])
  }
}

# Rename for clarity
df <- df %>%
  rename(
    war_damage = q02_01,
    disaster = q02_02,
    head_death = q02_03,
    unemployment = q02_04,
    business_failure = q02_05,
    family_conflict = q02_06,
    income_decline = q02_07,
    asset_loss = q02_08,
    illness_onset = q02_09,
    prolonged_illness = q02_10,
    aging_work_decline = q02_11,
    low_living_ability = q02_21,
    weak_household_head = q02_22,
    household_discord = q02_23,
    disabled_household_member = q02_24,
    longterm_patient = q02_25
  )

cat(sprintf("   Low-income-cause variables created: 16\n"))

# ============================================================================
# 7. INCOME AND CONSUMPTION VARIABLES
# ============================================================================

cat("\n7. Constructing income and consumption variables...\n")

df <- df %>%
  mutate(
    income_decreased = if_else(q08 == 3, 1L, if_else(q08 == 9, NA_integer_, 0L)),
    rent_arrears_amount = if_else((q19 < 1000 & q19 != 88888), q19, NA_real_),
    has_rent_arrears = if_else(rent_arrears_amount > 0, 1L, 0L)
  )

cat(sprintf("   Income decreased: %.1f%%\n", 100 * mean(df$income_decreased, na.rm = TRUE)))
cat(sprintf("   Has rent arrears: %.1f%%\n", 100 * mean(df$has_rent_arrears, na.rm = TRUE)))

# ============================================================================
# 8. HOUSEHOLD TYPE
# ============================================================================

cat("\n8. Household type...\n")
df <- df %>%
  mutate(household_type = q03)
cat(sprintf("   Household types: %d\n", n_distinct(df$household_type)))

# ============================================================================
# 9. SELECT AND SAVE ANALYSIS DATASET
# ============================================================================

cat("\n9. Saving analysis dataset...\n")

# Select key variables
df_analysis <- df %>%
  select(
    ID, city, household_type,
    # Coping strategies
    coping_account, coping_pawn, coping_employer, coping_friend,
    coping_asset_sale, coping_savings, coping_other, coping_food,
    any_coping, n_coping_strategies, any_borrowing,
    # Institutions
    any_welfare_loan, public_pawnshop, public_assistance,
    both_institutions, welfare_only, pawnshop_only, neither_institution,
    # Demographics
    head_age, head_age_sq, female_head, household_size, n_workers, n_unemployed,
    # Assets
    asset_count, asset_count_with_realestate,
    # Risk factors
    war_damage, disaster, head_death, unemployment, business_failure,
    family_conflict, income_decline, asset_loss, illness_onset,
    prolonged_illness, aging_work_decline, low_living_ability,
    weak_household_head, household_discord, disabled_household_member,
    longterm_patient,
    # Income
    income_decreased, has_rent_arrears, rent_arrears_amount,
    # Original variables (for reference)
    starts_with("q04_"), starts_with("q10_"), q07, q08
  )

# Save to CSV
output_file <- file.path(data_derived, "ssjda1331_analysis.csv")
write_csv(df_analysis, output_file)
cat(sprintf("   Saved to: %s\n", output_file))
cat(sprintf("   Dimensions: %d x %d\n", nrow(df_analysis), ncol(df_analysis)))

# ============================================================================
# 10. DATA QUALITY SUMMARY
# ============================================================================

cat("\n10. Data quality summary...\n")
cat("\n   Missing values (key institutional variables):\n")
cat(sprintf("     any_welfare_loan: %d\n", sum(is.na(df_analysis$any_welfare_loan))))
cat(sprintf("     public_pawnshop: %d\n", sum(is.na(df_analysis$public_pawnshop))))

cat("\n   Missing values (key demographic variables):\n")
cat(sprintf("     head_age: %d\n", sum(is.na(df_analysis$head_age))))
cat(sprintf("     female_head: %d\n", sum(is.na(df_analysis$female_head))))
cat(sprintf("     n_workers: %d\n", sum(is.na(df_analysis$n_workers))))

cat("\n   Coping variables: Complete for all observations\n")

cat("\n================================================================================\n")
cat("DATA CLEANING COMPLETE (R)\n")
cat("================================================================================\n")
cat(sprintf("Output file: %s\n", output_file))
cat("Ready for regression analysis.\n")

# Save environment for next script
save.image(file.path(output_logs, "phase1_workspace.RData"))
cat(sprintf("Workspace saved to: %s\n", file.path(output_logs, "phase1_workspace.RData")))
